drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
+        span: FlashRange,
+    },
+    Ram {
+        buffer: &'a [u8],
+        span: FlashRange,
+    },
+    // Hybrid is used for later implementation of rollback protection.
+    // The buffer is used in place of the beginning of the flash range.
+    _Hybrid {
+        buffer: &'a [u8],
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+}
+
+impl Accessor<'_> {
+    fn exec(&self) -> &Range<u32> {
+        match self {
+            Accessor::Flash { flash: _, span }
+            | Accessor::Ram { buffer: _, span }
+            | Accessor::_Hybrid {
+                buffer: _,
+                flash: _,
+                span,
+            } => &span.exec,
+        }
+    }
+}
+
+/// The update_server needs to deal with images that are:
+///   - Entirely in flash
+///   - Header block in RAM with the remainder unavailable
+///   - Header block in RAM with the remainder in flash.
+///   - Entire image in RAM (cached Stage0)
+/// Calls to methods use image relative addresses.
+/// ImageAccess::Flash{_, span} gives the physical flash addresses
+/// being accessed.
+///
+/// Any address here is the "storage" address which for stage0next
+/// and any image held temporarily in RAM is different than the
+/// execution address.
+pub struct ImageAccess<'a> {
+    accessor: Accessor<'a>,
+}
+
+impl ImageAccess<'_> {
+    pub fn new_flash<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Flash { flash, span },
+        }
+    }
+
+    pub fn new_ram(
+        buffer: &[u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'_> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Ram { buffer, span },
+        }
+    }
+
+    pub fn _new_hybrid<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        buffer: &'a [u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::_Hybrid {
+                flash,
+                buffer,
+                span,
+            },
+        }
+    }
+
+    fn exec(&self) -> &Range<u32> {
+        self.accessor.exec()
+    }
+
+    /// True if the u32 at offset is contained within the image.
+    pub fn is_addressable(&self, offset: u32) -> bool {
+        let len = match &self.accessor {
+            Accessor::Flash { flash: _, span }
+            | Accessor::_Hybrid {
+                buffer: _,
+                flash: _,
+                span,
+            }
+            | Accessor::Ram { buffer: _, span } => {
+                span.store.end.saturating_sub(span.store.start)
+            }
+        };
+        offset < len && offset.saturating_add(U32_SIZE) <= len
```

Comment:
So, I'm increasingly convinced as I read through this that every use of saturating arithmetic in this module should be replaced by checked. This does _not_ test the condition the comments suggest in the edge case where the integers could overflow.

Reply:
I'm making that change which makes it easier to reason about the code.